### ISSUE TITLE
New popup displayed behind old one

### ISSUE THREAD

**Issue by kaddkaka** [neovim/neovim#31543]
### Problem

When a new popup is displayed, there seems to be a risk that it is shown behind an older popup. The reverse ("zindex") order would be better so that the newest popup is fully visible.

### Steps to reproduce

For example, using `vim.lsp.buf.hover()` to spawn a popup, and then inside that popup calling `<Cmd>lua vim.diagnostic.goto_next({float={source=true}})<CR>` causes the diagnostics popup to appear below/behind the hover popup.

![Image](https://github.com/user-attachments/assets/de41653b-e7da-40b5-9b05-4155d9660700)

### Expected behavior

Newest popup to appear on top

### Nvim version (nvim -v)

v0.10.2

### Vim (not Nvim) behaves the same?

dont know

### Operating system/version

Ubuntu 24.04

### Terminal name/version

Wezterm

### $TERM environment variable

xterm-256color

### Installation

dont remember

**Comment by zeertzjq** [comment]
> For example, using `vim.lsp.buf.hover()` to spawn a popup, and then inside that popup calling `<Cmd>lua vim.diagnostic.goto_next({float={source=true}})<CR>` causes the diagnostics popup to appear below/behind the hover popup.

How can you do that? There are no diagnostics in the LSP popup window.

**Comment by luukvbaal** [comment]
> How can you have diagnostics in the LSP popup window?

Does it matter? I think it makes sense if the compositor would place child/newer windows on top of parent/older windows. Cycling through the windows with `Ctrl-W` in this repro should show the problem:

```lua
vim.cmd.call("setline(1, ['foobar']->repeat(100))")
vim.api.nvim_open_win(0, true, { relative = 'editor', row = 1, col = 1, width = 10, height = 5 })
vim.api.nvim_open_win(vim.api.nvim_create_buf(false, false), true, { relative = 'editor', row = 3, col = 3, width = 10, height = 5 })
vim.cmd.call("setline(1, ['newest']->repeat(100))")
```

If we don't fix the compositor I suppose we could make sure our runtime does not open windows with equal zindex to parent windows.

**Comment by zeertzjq** [comment]
> > How can you have diagnostics in the LSP popup window?
> 
> Does it matter? I think it makes sense if the compositor would place child/newer windows on top of parent/older windows. Cycling through the windows with `Ctrl-W` in this repro should show the problem:
> 
> vim.cmd.call("setline(1, ['foobar']->repeat(100))")
> vim.api.nvim_open_win(0, true, { relative = 'editor', row = 1, col = 1, width = 10, height = 5 })
> vim.api.nvim_open_win(vim.api.nvim_create_buf(false, false), true, { relative = 'editor', row = 3, col = 3, width = 10, height = 5 })
> vim.cmd.call("setline(1, ['newest']->repeat(100))")
> 
> If we don't fix the compositor I suppose we could make sure our runtime does not open windows with equal zindex to parent windows.

That may not be the same problem. In this script the two floating windows are created in the same event loop iteration, but in OP they aren't.

**Comment by zeertzjq** [comment]
Ah, that seems to be intentional. It happened here:
```c
  // TODO(bfredl): maybe not the best time to do this, for efficiency we
  // should configure all grids before entering win_update()
  if (curgrid != &default_grid) {
    size_t new_index = kv_size(layers) - 1;

    while (new_index > 1 && kv_A(layers, new_index)->zindex > curgrid->zindex) {
      new_index--;
    }

    if (curgrid->comp_index < new_index) {
      ui_comp_raise_grid(curgrid, new_index);
    }
  }
```

**Comment by luukvbaal** [comment]
Yeah there's something for the curgrid behavior. Then maybe it should be up to the caller to choose an appropriate zindex.

**Comment by kaddkaka** [comment]
> > For example, using `vim.lsp.buf.hover()` to spawn a popup, and then inside that popup calling `<Cmd>lua vim.diagnostic.goto_next({float={source=true}})<CR>` causes the diagnostics popup to appear below/behind the hover popup.
> 
> How can you have diagnostics in the LSP popup window?

By using otter.nvim: https://github.com/jmbuhr/otter.nvim/issues/189

**Comment by kaddkaka** [comment]
So what would be the way to fix this? It seems like a potential minor change to fix this (If it's just modifying a zindex value update  somewhere)
